### PR TITLE
query param with better UX

### PR DESCRIPTION
**Pull request by rohanpai**

### CODE REVIEW

File: app/web_modules/sourcegraph/search/SearchSettings.js [hunk 1]
```diff
@@ -97,6 +119,7 @@ class SearchSettings extends Container {
 
 		Dispatcher.Stores.dispatch(new UserActions.UpdateSettings(newSettings));
 
+
```
Comment (rohanpai) [inline app/web_modules/sourcegraph/search/SearchSettings.js]: will remove this

File: app/web_modules/sourcegraph/app/GlobalNav.js [hunk 2]
```diff
@@ -221,7 +222,9 @@ class SearchForm extends React.Component {
 
 	_handleSubmit(ev: Event) {
 		ev.preventDefault();
-		this.props.router.push(locationForSearch(this.props.location, this.state.query, true, true));
+		const settings = UserStore.settings.get();
+		const langs = settings && settings.search ? settings.search.languages : null;
```
Comment (rothfels) [inline app/web_modules/sourcegraph/app/GlobalNav.js]: Include the "repos to include" filter in query parameters as well.

File: app/web_modules/sourcegraph/search/SearchSettings.js [hunk 3]
```diff
@@ -35,10 +34,33 @@ class SearchSettings extends Container {
...
+	}
+
+	componentDidMount() {
+		super.componentDidMount();
+		const langFromQuery = this.props.location.query.lang;
+		if (langFromQuery) {
+			const langs = (typeof (langFromQuery) === "string") ? [langFromQuery] : langFromQuery;
+			let validLangParams = [];
+			for (const lang in langs) {
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/SearchSettings.js]: Instead of lines 49 - 53, you can make `langs` a `let` instead of a `const` and then do the following:

`langs = langs.filter(langIsSupported)`

^^ much nicer :)

File: app/web_modules/sourcegraph/search/routes.js [hunk 4]
```diff
@@ -29,23 +29,32 @@ export function urlToSearch(query?: ?string): string {
 // location and the updatePath/updateQuery parameters.
 //
 // This function modifies loc.
-export function locationForSearch(loc: RouterLocation, query: ?string, updatePath: bool, updateQuery: bool): RouterLocation {
+export function locationForSearch(loc: RouterLocation, query: ?string, lang: ?string[], updatePath: bool, updateQuery: bool): RouterLocation {
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: s/lang/langs

File: app/web_modules/sourcegraph/search/routes.js [hunk 5]
```diff
@@ -29,23 +29,32 @@ export function urlToSearch(query?: ?string): string {
...
 	if (updatePath) {
 		loc.pathname = `/${rel.search}`;
 	}
+	let langParam = [];
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: I think your flow type annotation makes this irrelevant; langs should always be an array.

File: app/web_modules/sourcegraph/search/routes.js (hunk 4 above)
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: Maybe would be better (esp. since you'd be adding a third param with the "repos" filter) to just make this one parameter called `searchParams` which is an object w/ `query`, `langs`, and `repos`.

File: app/web_modules/sourcegraph/search/GlobalSearch.js [hunk 6]
```diff
@@ -169,9 +171,9 @@ class GlobalSearch extends Container {
 			} else {
 				state._privateRepos = null;
 			}
+			this.context.router.push(locationForSearch(this.props.location, this.state.query, languages, scope, true, true));
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/GlobalSearch.js]: This should be a replace i think

File: app/web_modules/sourcegraph/search/routes.js [hunk 7]
```diff
@@ -29,23 +29,42 @@ export function urlToSearch(query?: ?string): string {
...
 // This function modifies loc.
-export function locationForSearch(loc: RouterLocation, query: ?string, updatePath: bool, updateQuery: bool): RouterLocation {
+export function locationForSearch(loc: RouterLocation, query: ?string, lang: ?string[], scope: ?Object, updatePath: bool, updateQuery: bool): RouterLocation {
 	if (updatePath) {
 		loc.pathname = `/${rel.search}`;
 	}
+	let langParam = [];
+	if (typeof (lang) === "string") {
+		langParam = [lang];
+	} else if (typeof (lang) === "object" && lang !== null && lang.length > 0) {
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: This stuff is too complicated. Why isn't lang always a string array?

File: app/web_modules/sourcegraph/search/routes.js [hunk 8]
```diff
@@ -29,23 +29,42 @@ export function urlToSearch(query?: ?string): string {
...
 
 	if (!updateQuery) {
 		if (query && !loc.state) loc.state = {};
-		if (query) loc.state.q = query;
-		else delete loc.state.q;
+		if (query) {
+			loc.state.q = query;
+			loc.state.lang = langParam;
+			scopeNames.map(function(scopeName) {
+				if (scope[scopeName] === true) loc.state[scopeName] = true;
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: Instead of accessing these properties directly on `loc.state`, you should use a `loc.state.scope` object.

File: app/web_modules/sourcegraph/search/routes.js [hunk 9]
```diff
@@ -29,23 +29,42 @@ export function urlToSearch(query?: ?string): string {
...
 	}
 
 	if (query && !loc.query) loc.query = {};
-	if (query) loc.query.q = query;
-	else delete loc.query.q;
+	if (query) {
+		loc.query.q = query;
+		loc.query.lang = langParam;
+		scopeNames.map(function(scopeName) {
+			if (scope[scopeName] === true) loc.query[scopeName] = true;
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: I see duplicated code here. Factor.

File: app/web_modules/sourcegraph/search/SearchSettings.js [hunk 10]
```diff
@@ -35,10 +34,38 @@ class SearchSettings extends Container {
...
+		super(props);
+		this.state = {settings: UserStore.settings.get()};
+
+	}
+
+	componentDidMount() {
+		super.componentDidMount();
+		const langFromQuery = this.props.location.query.lang;
+		if (langFromQuery) {
+			const langs = (typeof (langFromQuery) === "string") ? [langFromQuery] : langFromQuery;
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/SearchSettings.js]: Remove parens around `langFromQuery`

Comment (rothfels) [inline app/web_modules/sourcegraph/search/SearchSettings.js]: You can also remove them around the `typeof` clause

File: app/web_modules/sourcegraph/search/SearchSettings.js [hunk 11]
```diff
@@ -35,10 +34,38 @@ class SearchSettings extends Container {
...
+				...this.state.settings,
+				search: {
+					...this.state.settings.search,
+					languages: validLangParams,
+					scope: scopeParams,
+				},
+			};
+			setTimeout(() => Dispatcher.Stores.dispatch(new UserActions.UpdateSettings(newSettings), 3000));
+		} else {
+			setTimeout(() => Dispatcher.Stores.dispatch(new UserActions.UpdateSettings(this.state.settings), 3000));
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/SearchSettings.js]: You shouldn't need to do this. Why is this here?

File: app/web_modules/sourcegraph/search/SearchSettings.js [hunk 12]
```diff
@@ -35,10 +34,38 @@ class SearchSettings extends Container {
...
+
+			const newSettings = {
+				...this.state.settings,
+				search: {
+					...this.state.settings.search,
+					languages: validLangParams,
+					scope: scopeParams,
+				},
+			};
+			setTimeout(() => Dispatcher.Stores.dispatch(new UserActions.UpdateSettings(newSettings), 3000));
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/SearchSettings.js]: Why such a long timeout? You should also leave a helpful comment so the next developer understands why this is here.

File: app/web_modules/sourcegraph/search/routes.js [hunk 13]
```diff
@@ -29,26 +29,42 @@ export function urlToSearch(query?: ?string): string {
...
 
+	if (!lang) lang = [];
+
 	if (!updateQuery) {
+		loc.state = updateScopeAndLanguage(loc.state, scope, lang);
 		if (query && !loc.state) loc.state = {};
-		if (query) loc.state.q = query;
-		else delete loc.state.q;
+		if (query) {
+			loc.state.q = query;
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: If you're going to inline the `else` condition, you can also inline this condition.

File: app/web_modules/sourcegraph/search/routes.js [hunk 14]
```diff
@@ -29,26 +29,42 @@ export function urlToSearch(query?: ?string): string {
...
+		} else delete loc.state.q;
 		return loc;
 	}
 
+	loc.query = updateScopeAndLanguage(loc.query, scope, lang);
 	if (query && !loc.query) loc.query = {};
-	if (query) loc.query.q = query;
-	else delete loc.query.q;
+	if (query) {
+		loc.query.q = query;
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: Et al.

File: app/web_modules/sourcegraph/search/routes.js [hunk 15]
```diff
@@ -29,26 +29,42 @@ export function urlToSearch(query?: ?string): string {
...
 }
 
+function updateScopeAndLanguage(oldState: any, scope, lang) {
+	const scopeNames = ["public", "private", "popular", "repo"];
+	let state = {...oldState};
+	scopeNames.map((scopeName) => {
+		if (scope && scope[scopeName]) state[scopeName] = true;
+		else delete state[scopeName];
+	});
+	state.lang = lang.slice();
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: What is this line for?

File: app/web_modules/sourcegraph/search/routes.js [hunk 16]
```diff
@@ -29,26 +29,42 @@ export function urlToSearch(query?: ?string): string {
...
+	if (query) {
+		loc.query.q = query;
+	} else delete loc.query.q;
 	if (loc.state) delete loc.state.q;
-
 	return loc;
 }
 
+function updateScopeAndLanguage(oldState: any, scope, lang) {
+	const scopeNames = ["public", "private", "popular", "repo"];
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: You've hardcoded these here, and also have new code in SearchSettings.js that iteratively looks over these properties. Somebody is bound to update one of these locations but not the other, so you should extract these scope values into a common location and import them.

Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: I recommend `search/index.js` where the `SearchScope` type is defined.

File: app/web_modules/sourcegraph/search/routes.js (hunk 15 above)
Comment (rohanpai) [inline app/web_modules/sourcegraph/search/routes.js]: @rothfels I need this line since otherwise I get the error `Uncaught TypeError: Cannot assign to read only property '1' of object '[object Array]'`

File: app/web_modules/sourcegraph/search/SearchSettings.js [hunk 17]
```diff
@@ -11,15 +11,15 @@ import Dispatcher from "sourcegraph/Dispatcher";
...
-import {allLangs, langName} from "sourcegraph/Language";
+import {allLangs, langName, langIsSupported} from "sourcegraph/Language";
 import type {LanguageID} from "sourcegraph/Language";
 import {privateGitHubOAuthScopes} from "sourcegraph/util/urlTo";
 import {withUserContext} from "sourcegraph/app/user";
 import LocationStateToggleLink from "sourcegraph/components/LocationStateToggleLink";
 import {LocationStateModal, dismissModal} from "sourcegraph/components/Modal";
 import BetaInterestForm from "sourcegraph/home/BetaInterestForm";
 import * as AnalyticsConstants from "sourcegraph/util/constants/AnalyticsConstants";
+import {SearchScopes} from "sourcegraph/search";
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/SearchSettings.js]: This should not be capitalized; reserve that for components/classes. Use `searchScopes` camel case.

File: app/web_modules/sourcegraph/search/SearchSettings.js [hunk 18]
```diff
@@ -35,10 +35,35 @@ class SearchSettings extends Container {
...
+
+	componentDidMount() {
+		super.componentDidMount();
+		const langFromQuery = this.props.location.query.lang ? this.props.location.query.lang : [];
+		const query = this.props.location.query;
+		if (this.props.location.pathname === "/search" && (query.lang || query.public || query.private || query.repo || query.popular)) {
+			const langs = typeof langFromQuery === "string" ? [langFromQuery] : langFromQuery;
+			const validLangParams = langs.filter(langIsSupported);
+			let scopeParams = {};
+			SearchScopes.map(function(scopeName) {
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/SearchSettings.js]: only use `.map` when you care about the return value; else use `.forEach`

Also, if you use the `=>` syntax you can avoid `.bind(this)`:

```
let scopeParams = {};
searchScopes.forEach((scopeName) => scopeParams[scopeName] = this.props.locatino.query[scopeName] === "true");
```

File: app/web_modules/sourcegraph/search/SearchSettings.js [hunk 19]
```diff
@@ -35,10 +35,35 @@ class SearchSettings extends Container {
...
+
+			const newSettings = {
+				...this.state.settings,
+				search: {
+					...this.state.settings.search,
+					languages: validLangParams,
+					scope: scopeParams,
+				},
+			};
+			setTimeout(() => Dispatcher.Stores.dispatch(new UserActions.UpdateSettings(newSettings), 100));
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/SearchSettings.js]: Are you sure you need a timeout of 100ms even? It may be enough to exclude the second argument entirely (which just guarantees that the callback will be invoked after the component mount cycle completes.

File: app/web_modules/sourcegraph/search/routes.js [hunk 20]
```diff
@@ -29,26 +30,42 @@ export function urlToSearch(query?: ?string): string {
 // location and the updatePath/updateQuery parameters.
 //
 // This function modifies loc.
-export function locationForSearch(loc: RouterLocation, query: ?string, updatePath: bool, updateQuery: bool): RouterLocation {
+// @NOTE loc cannot have immutable properties. lang and scope might be immutable.
+//		 Ensure that if you assign new properties to loc that the properties are mutable
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: Remove leading whitespace before Ensure, and end sentences with period.

File: app/web_modules/sourcegraph/search/routes.js [hunk 21]
```diff
@@ -58,3 +75,19 @@ export function queryFromStateOrURL(loc: RouterLocation): ?string {
 	else if (loc.query && loc.query.hasOwnProperty("q")) return firstQueryValue(loc.query.q);
 	return null;
 }
+
+
+export function langFromStateOrURL(loc: RouterLocation): ?string[] {
+	if (loc.state && loc.state.hasOwnProperty("lang")) return typeof loc.state.lang === "string" ? [loc.state.lang] :loc.state.lang;
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: why not just `loc.state.lang`?

File: app/web_modules/sourcegraph/search/routes.js [hunk 22]
```diff
@@ -58,3 +75,19 @@ export function queryFromStateOrURL(loc: RouterLocation): ?string {
 	else if (loc.query && loc.query.hasOwnProperty("q")) return firstQueryValue(loc.query.q);
 	return null;
 }
+
+
+export function langFromStateOrURL(loc: RouterLocation): ?string[] {
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: Rename `langsFromStateOrURL`

File: app/web_modules/sourcegraph/search/routes.js [hunk 23]
```diff
@@ -58,3 +75,19 @@ export function queryFromStateOrURL(loc: RouterLocation): ?string {
...
+	if (loc.state && loc.state.hasOwnProperty("lang")) return typeof loc.state.lang === "string" ? [loc.state.lang] :loc.state.lang;
+	else if (loc.query && loc.query.hasOwnProperty("lang")) return typeof loc.query.lang === "string" ? [loc.query.lang] : loc.query.lang;
+	return null;
+}
+
+export function scopeFromStateOrURL(loc: RouterLocation): ?Object {
+	let scope = {};
+	searchScopes.forEach((scopeName) => {
+		if (loc.state && loc.state.hasOwnProperty(scopeName)) scope[scopeName] = loc.state.scopeName;
```
Comment (rothfels) [inline app/web_modules/sourcegraph/search/routes.js]: why not just `loc.state[scopeName]`

Comment (rohanpai) [inline app/web_modules/sourcegraph/search/routes.js]: loc.state might be undefined. 

Merging after passes CI tests in GH
